cmd/mesh/upgrade.go

```diff
@@ -0,0 +1,353 @@
+			"istioctl: %v\nIstio control plane: %v", cur, tar)
+	}
+
+	if curPkgVer.Patch > tarPkgVer.Patch {
+		l.logAndFatalf("Abort. A newer version has been installed in the cluster.\n"+
+			"istioctl: %v\nIstio control plane: %v", cur, tar)
+	}
+
+	tarGoVersion, err := goversion.NewVersion(tar)
+	if err != nil {
+		l.logAndFatalf("Abort. Failed to parse the target version: %v", tar)
+	}
+
+	compatibleMap := getVersionCompatibleMap(versionsURI, tarGoVersion, l)
+
+	curGoVersion, err := goversion.NewVersion(cur)
+	if err != nil {
+		l.logAndFatalf("Abort. Failed to parse the current version: %v", cur)
+	}
+
+	if !compatibleMap.SupportedIstioVersions.Check(curGoVersion) {
+		l.logAndFatalf("Abort. Upgrade is currently not supported: %v -> %v.", cur, tar)
+	}
+
+	l.logAndPrintf("Upgrade version check passed: %v -> %v.\n", cur, tar)
+}
+
+// retrieveControlPlaneVersion retrieves the version number from the Istio control plane
+func retrieveControlPlaneVersion(kubeClient manifest.ExecClient, istioNamespace string) string {
+	meshInfo, e := kubeClient.GetIstioVersions(istioNamespace)
+	if e != nil {
+		l.logAndFatalf("Failed to retrieve Istio control plane version, error: %v", e)
+	}
+
+	if meshInfo == nil {
+		l.logAndFatalf("Istio control plane not found in namespace: %v", istioNamespace)
+	}
+
+	for _, remote := range *meshInfo {
+		l.logAndPrintf("Control Plane - %s pod - version: %s", remote.Component, remote.Info.Version)
+	}
+	l.logAndPrint("")
+
+	return coalesceVersions(meshInfo)
+}
+
+// waitUpgradeComplete waits for the upgrade to complete by periodically comparing the current component version
+// to the target version.
+func waitUpgradeComplete(kubeClient manifest.ExecClient, istioNamespace string, targetVer string) {
+	for i := 1; i <= upgradeWaitCheckVerMaxAttempts; i++ {
+		l.logAndPrintf("Waiting for upgrade rollout to complete, attempt #%v: ", i)
+		sleepSeconds(upgradeWaitSecCheckVerPerLoop)
+		meshInfo, e := kubeClient.GetIstioVersions(istioNamespace)
+		if e != nil {
+			l.logAndPrintf("Failed to retrieve Istio control plane version, error: %v", e)
+			continue
+		}
+		if meshInfo == nil {
+			l.logAndPrintf("Failed to find Istio namespace: %v", istioNamespace)
+			continue
+		}
+		if identicalVersions(*meshInfo) && targetVer == (*meshInfo)[0].Info.Version {
+			l.logAndPrintf("Upgrade rollout completed. " +
+				"All Istio control plane pods are running on the target version.\n\n")
+			return
+		}
+		for _, remote := range *meshInfo {
+			if targetVer != remote.Info.Version {
+				l.logAndPrintf("Control Plane - %s pod - version %s does not match the target version %s",
+					remote.Component, remote.Info.Version, targetVer)
+			}
+		}
+	}
+	l.logAndFatal("Upgrade rollout unfinished. Maximum number of attempts exceeded, quit...")
+}
+
+// sleepSeconds sleeps for n seconds, printing a dot '.' per second
+func sleepSeconds(n int) {
+	l.logAndPrintf("Going to sleep for %v secounds", n)
```

Comment:
i really don't think we need the above log message - keep in mind these are for users, not developers. it's not something that a user would expect to see in a tool.

Reply:
Yes, removed.